src/utils/openmp_wrapper.cpp

```diff
@@ -0,0 +1,44 @@
+/*!
+ * Copyright (c) 2023 Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See LICENSE file in the project root for license information.
+ */
+#include <LightGBM/utils/openmp_wrapper.h>
+
+int LGBM_MAX_NUM_THREADS = -1;
+
+int LGBM_DEFAULT_NUM_THREADS = -1;
+
+#ifdef _OPENMP
+
+#include <omp.h>
+
+int OMP_NUM_THREADS() {
+  int default_num_threads = 1;
+
+  if (LGBM_DEFAULT_NUM_THREADS > 0) {
+    // if LightGBM-specific default has been set, ignore OpenMP-global config
+    default_num_threads = LGBM_DEFAULT_NUM_THREADS;
+  } else {
+    // otherwise, default to OpenMP-global config
+    #pragma omp single
+    { default_num_threads = omp_get_max_threads(); }
```

Comment:
Wouldn't this make the python scikit-learn interface non-compliant with what it currently says in the docs ("Negative integers are interpreted as following joblib’s formula") when the user happens to have a different default configured for openmp?

Reply:
Actually no, it wouldn't, since the calculation from joblib's formula is performed explicitly inside the scikit-learn interface at both `fit` and `predict`.

Reply:
Yep exactly, thanks to your changes from #5105, `num_threads = -1` shouldn't reach this code path for the `scikit-learn` interface.